Fix dgst output for HMAC

mattcaswell opened the pull request:
The dgst app was using an undocumented behaviour in the
EVP_PKEY_new_raw_private_key() function when setting a key length for
a MAC. The old EVP_PKEY to MAC bridge, probably by accident, converts a
-1 length to a strlen() call, by virtue of the fact that it eventually
calls ASN1_STRING_set() which has this feature.

As noted above this is undocumented, and unexpected since the len
parameter to EVP_PKEY_new_raw_private_key() is an unsigned value (size_t).
In the old bridge it was later (silently) cast to an int, and therefore
the original -1 value was restored. This only works because sizeof(int) <=
sizeof(size_t). If we ever run on a platform where sizeof(int) >
sizeof(size_t) then it would have failed. The behaviour also doesn't hold
for EVP_PKEY_new_raw_private_key() in general - only when the old MAC
bridge was in use.

Rather than restore the original behaviour I think it is best to simply
fix the dgst app to not assume it exists. We should not bake in this
backwards and inconsistent behaviour.

Fixes #12837

t8m reviewed (approved):
It would be good if Dmitry confirmed that it fixes the regression issue.

beldmit commented:
Don't hurry :)

beldmit commented:
Unfortunately, it fixes the problem only partially.

1. Minor nit: previously I got output containing word HMAC-mdname(filename), now I get a misleading mdname(filename). I can fix the test if it is the desired behavior.

2. I get a new failure. When I call 
```
./openssl_wrap.sh dgst -md_gost12_512 -hmac 123456901234567890123456789012 somefile somefile
```
I get different values for the files.

3. Last but not least, I'll check it myself again though it may be the bug similar to previous.
I got a segfault executing the command 
./openssl_wrap.sh dgst -md_gost94 -hmac 123456901234567890123456789012 tcl_tests/manul-gost/dgst.dat tcl_tests/manul-gost/dgst.dat
```
#0  __memset_sse2_unaligned () at ../sysdeps/x86_64/multiarch/memset-vec-unaligned-erms.S:120
#1  0x00007ffff7e08f35 in drbg_ctr_generate (drbg=0x5555556affc0, out=0x20 <error: Cannot access memory at address 0x20>, outlen=32, adin=0x0, adinlen=0) at providers/implementations/rands/drbg_ctr.c:406
#2  0x00007ffff7e1256f in PROV_DRBG_generate (drbg=0x5555556affc0, out=0x20 <error: Cannot access memory at address 0x20>, outlen=32, strength=<optimized out>, prediction_resistance=0, adin=0x0, adinlen=0)
    at providers/implementations/rands/drbg.c:708
#3  0x00007ffff7cf21b8 in evp_rand_generate_locked (ctx=ctx@entry=0x55555575b950, out=out@entry=0x20 <error: Cannot access memory at address 0x20>, outlen=outlen@entry=32, strength=strength@entry=0, 
    prediction_resistance=<optimized out>, addin=0x0, addin_len=0) at crypto/evp/evp_rand.c:495
#4  0x00007ffff7cf2e59 in EVP_RAND_generate (ctx=0x55555575b950, out=out@entry=0x20 <error: Cannot access memory at address 0x20>, outlen=outlen@entry=32, strength=strength@entry=0, prediction_resistance=prediction_resistance@entry=0, 
    addin=addin@entry=0x0, addin_len=0) at crypto/evp/evp_rand.c:517
#5  0x00007ffff7d4d9b1 in RAND_priv_bytes_ex (ctx=0x0, buf=0x20 <error: Cannot access memory at address 0x20>, num=32) at crypto/rand/rand_lib.c:279
#6  0x00007ffff7853968 in gost_key (c=0x0, k=0x7fffffffd9b0 "Taete s sttdto.a  \ntddTaaie tgst") at gost89.c:455
#7  0x00007ffff7853933 in gost_enc_with_key (c=0x0, key=0x0, inblock=0x555555740594 "", outblock=0x7fffffffd9f0 "\020\204dUUU") at gost89.c:447
#8  0x00007ffff7854b4d in hash_step (c=0x0, H=0x555555740594 "", 
    M=0x55555565f970 "Test data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data t"...)
    at gosthash.c:97
#9  0x00007ffff785495f in hash_block (ctx=0x555555740580, 
    block=0x55555565f970 "Test data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data t"..., length=2100) at gosthash.c:222
#10 0x00007ffff7cd6dce in md_read (outl=<optimized out>, 
    out=0x55555565f970 "Test data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data t"..., b=0x555555667790) at crypto/evp/bio_md.c:93
#11 md_read (b=0x555555667790, 
    out=0x55555565f970 "Test data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data t"..., outl=<optimized out>) at crypto/evp/bio_md.c:75
#12 0x00007ffff7c08f0a in bread_conv (bio=<optimized out>, data=<optimized out>, datal=<optimized out>, readbytes=0x7fffffffdaf8) at crypto/bio/bio_meth.c:123
#13 0x00007ffff7c076b5 in bio_read_intern (readbytes=0x7fffffffdaf8, dlen=8192, data=0x55555565f970, b=0x555555667790) at crypto/bio/bio_lib.c:270
#14 BIO_read (b=b@entry=0x555555667790, data=data@entry=0x55555565f970, dlen=dlen@entry=8192) at crypto/bio/bio_lib.c:296
#15 0x00005555555a8c6c in do_fp (out=out@entry=0x555555667860, 
    buf=buf@entry=0x55555565f970 "Test data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data to digest.\nTest data t"..., bp=bp@entry=0x555555667790, sep=sep@entry=0, binout=binout@entry=0, key=0x5555556ae890, sigin=0x0, siglen=0, sig_name=0x0, md_name=0x7ffff7e49cbd "md_gost94", file=0x7fffffffe27c "tcl_tests/manul-gost/dgst.dat")
    at apps/dgst.c:524
#16 0x00005555555a9ade in dgst_main (argc=2, argv=0x7fffffffdea0) at apps/dgst.c:427
#17 0x00005555555b7dfb in do_cmd (prog=0x5555556486b0, argc=6, argv=0x7fffffffde80) at apps/openssl.c:406
#18 0x00005555555952c9 in main (argc=6, argv=0x7fffffffde80) at apps/openssl.c:287

```

beldmit commented:
As (2) seems to be a bug, I take off the approval label.

beldmit commented:
There is a lack of either reinitialization of a MAC context or duplication of it.

paulidale commented:
> There is a lack of either reinitialization of a MAC context or duplication of it.

There was a discussion in #8363 where this was not wanted.

beldmit commented:
If we want to avoid changing the `openssl dgst -hmac` behavior when many files are passed, we have to either create a ctx for each file or create it once and duplicate it for each file.

beldmit commented:
As the @mattcaswell's fix makes the situation with fetching better - should I approve this fix and create a new issue related to HMACing multiple files which is a different beast?

mattcaswell commented:
I have fixes for the additional issues that  @beldmit identified. I just need a bit more time to get it into a reviewable state.

mattcaswell commented:
I've reworked this PR significantly and added several new commits addressing issues (1) and (2) above. I've not looked at (3) - probably that should be a different PR.

Please take another look.

mattcaswell commented:
...and of course it immediately needed a rebase which I've now done. Also I forgot to mention I renamed this PR to more accurately reflect the latest changes.

beldmit reviewed (approved):
LGTM

beldmit commented:
The (3) has gone too. Many thanks!

t8m commented on doc/man3/EVP_PKEY_is_a.pod:
```diff
@@ -28,6 +36,8 @@ otherwise 0.
 EVP_PKEY_can_sign() returns 1 if the I<pkey> key type functionality
 supports signing, otherwise 0.
 
+EVP_KEYMGMT_get0_first_name() returns the name that is found or NULL on error.
```
Should be EVP_PKEY_get0_first_alg_name().

levitte replied:
Interestingly enough, I'm adding `EVP_PKEY_typenames_do_all()` in #12873...

t8m replied:
This trivial doc copy&paste error is still unresolved.

t8m reviewed (changes requested):
One copy&paste error, otherwise LGTM

beldmit commented:
Travis failures seem relevant. Though I can't get why they happen on these 2 platforms...

levitte commented on crypto/evp/digest.c:
```diff
@@ -140,6 +140,25 @@ int EVP_DigestInit_ex(EVP_MD_CTX *ctx, const EVP_MD *type, ENGINE *impl)
+         * previously initialised with EVP_DigestSignInit() would retain
+         * information about the key, and re-initialise for another sign
+         * operation. So in that case we redirect to EVP_DigestSignInit()
+         */
```
Eh gods...

levitte commented on apps/dgst.c:
```diff
@@ -405,13 +406,8 @@ int dgst_main(int argc, char **argv)
-                                            NULL, NULL, &sig_name, ameth);
-            }
+            if (sigkey != NULL)
+                sig_name = EVP_PKEY_get0_first_alg_name(sigkey);
```
This is very dodgy.  One alias to the HMAC signature algorithm and the test may randomly go **BAMF**, because you might end up with, oh I don't know, "id-Hmac".  This is a case where the application has more accurate knowledge than libcrypto, as it really has *all* the data it needs to know exactly what names it's asking for.  It's quite a simple program after all...

mattcaswell replied:
Err, no. We're using it for HMAC here, but in theory you can sign with any type of key - so we really have to get the name from the key.

beldmit commented:
Removed approval labels because of Travis failures. Will try not to set it until Travis is happy.

mattcaswell commented:
> Travis failures seem relevant. Though I can't get why they happen on these 2 platforms...

Me too! I've added a bit more diagnostic information to see if that gives us any clues.

beldmit commented:
It does... Everything is OK with HMAC but weirdness is related to file paths.

mattcaswell commented:
Fixup pushed. Lets see if travis like this better.

beldmit commented:
Travis is happy.

beldmit reviewed (approved):
LGTM

mattcaswell commented:
I'm taking this PR out of the "beta1" milestone. This is just a bug fix. It will almost certainly be in before beta1, but it wouldn't have to be for beta 1 to go out.

t8m reviewed (changes requested):
Please fix the doc error. No need to reset the 24hour grace period after that.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

beldmit commented:
Merged fixing typo.
Many thanks!
